File: apilib/src/main/java/com/guider/health/apilib/IGuiderApi.java
package com.guider.health.apilib;

import com.guider.health.apilib.model.BeanOfWecaht;
import com.guider.health.apilib.model.CardInfo;
import com.guider.health.apilib.model.Devices;
import com.guider.health.apilib.model.DoctorInfo;
import com.guider.health.apilib.model.HasWechatId;
import com.guider.health.apilib.model.HealthAdvice;
import com.guider.health.apilib.model.IsFocusWechat;
import com.guider.health.apilib.model.ParamThirdUserAccount;
import com.guider.health.apilib.model.SimpleUserInfo;
import com.guider.health.apilib.model.TokenInfo;
import com.guider.health.apilib.model.UserInfo;
import com.guider.health.apilib.model.WechatInfo;
import com.guider.health.apilib.model.WechatUserInfo;
import com.guider.health.apilib.model.doctor.City;
import com.guider.health.apilib.model.doctor.JiGou;
import com.guider.health.apilib.model.doctor.MinZu;

import java.util.List;

import okhttp3.MultipartBody;
import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.DELETE;
import retrofit2.http.GET;
import retrofit2.http.Multipart;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Part;
import retrofit2.http.Path;
import retrofit2.http.Query;

public interface IGuiderApi {
    @GET("api/v1/healthadvice/page")
    Call<List<HealthAdvice>> getHealthAdvice(@Query("accountId") long accountId, @Query("page") int page, @Query("row") int row);

    @Multipart
    @POST("upload/file")
    Call<String> uploadFile(@Part MultipartBody.Part requestBody);

    @POST("api/v2/wechat/scan/login/userinfo")
    Call<BeanOfWecaht> getWechatInfo(@Query("doctorAccountId") long doctorAccountId,
                                     @Query("deviceCode") String deviceCode,
                                     @Query("wechatAppId") String wechatAppId,
                                     @Query("sence") String sence);

    @POST("api/v2/wechat/scan/login/tokeninfo")
    Call<BeanOfWecaht> getWechatTocken(@Query("doctorAccountId") long doctorAccountId,
                                       @Query("deviceCode") String deviceCode,
                                       @Query("wechatAppId") String wechatAppId,
                                       @Query("sence") String sence);

    /**
     * 获取当前PAD所属机构的AppId
     * @param mac
     * @return
     */
    @GET("api/v1/device/{mac}/wechatappid")
    Call<String> getWechatAppId(@Path("mac") String mac);

    /**
     * 微信登录
     *
     *                          JSONObject jsonObject = new JSONObject();
     *                         //HashMap<String, Object> request = new HashMap<>();
     *                         jsonObject.put("appId", "wx1971c8a28a09ad4b");
     *                         jsonObject.put("headimgurl", data.get("iconurl"));
     *                         jsonObject.put("nickname", data.get("name"));
     *                         jsonObject.put("openid", data.get("openid"));
     *
     *                         int gender = data.get("gender").equals("男") ? 0 : 1;
     *                         jsonObject.put("sex", gender);
     *                         jsonObject.put("unionid", data.get("unionid"));
     */
    @POST("api/v2/third/login/wachat/tokeninfo")
    Call<BeanOfWecaht> wechatLoginToken(@Body() WechatInfo paramWechatUserInfo);


    /**
     * 在需要关注微信公众号的时候轮询监测用户是否已经关注了公众号
     * @param sence
     * @return
     */
    @POST("api/v2/wechat/scan/group/wechatBind")
    Call<IsFocusWechat> checkIsFocus(@Query("sence") String sence);

    /**
     * 微信绑定群组
     *
     * @param sence
     * @param deviceCode
     * @return
     */
    @POST("api/v2/wechat/scan/group/userinfo")
    Call<BeanOfWecaht> getWechatGroup(@Query("sence") String sence,
                                      @Query("deviceCode") String deviceCode,
                                      @Query("wechatAppId") String wechatAppId
    );

    /**
     * 微信绑定手机号
     *
     * @param phone
     * @param code
     * @param sence
     * @return
     */
    @POST("api/v2/wechat/bind/phone/userinfo")
    Call<UserInfo> wechatBindPhone(@Query("phone") String phone,
                                   @Query("code") String code,
                                   @Query("deviceCode") String deviceCode,
                                   @Query("doctorAccountId") String doctorAccountId,
                                   @Query("wechatAppId") String wechatAppId,
                                   @Body() WechatUserInfo sence);

    /**
     * 身份证登录
     *
     * @return
     */
    @POST("api/v2/login/cardid/userinfo")
    Call<BeanOfWecaht> cardLogin(@Body() CardInfo cardInfo);

    /**
     * 身份证绑定手机号
     *
     * @param phone
     * @param code
     * @return
     */
    @POST("api/v2/cardid/bind/phone/userinfo")
    Call<UserInfo> cardBindPhone(@Query("phone") String phone,
                                 @Query("code") String code,
                                 @Query("cardId") String cardId,
                                 @Query("deviceCode") String deviceCode,
                                 @Query("doctorAccountId") String doctorAccountId,
                                 @Body() CardInfo cardInfo);

    /**
     * 微信绑定手机号
     *
     * @param phone
     * @param code
     * @param sence
     * @return
     */
    @POST("api/v2/wechat/bind/phone/token")
    Call<TokenInfo> wechatRegister(@Query("phone") String phone,
                                   @Query("code") String code,
                                   @Query("wechatAppId") String wechatAppId,
                                   @Body() WechatUserInfo sence);

    /**
     * 完善用户信息
     *
     * @param userInfo
     * @return
     */
    @PUT("api/v1/usersimpleinfo")
    Call<String> simpUserInfo(@Body() UserInfo userInfo);
    @PUT("api/v1/usersimpleinfo")
    Call<SimpleUserInfo> editSimpUserInfo(@Body() SimpleUserInfo userInfo);
    /**
     * 获取用户信息
     *
     * @return
     */
    @GET("api/v1/userinfo")
    Call<UserInfo> getUserInfo(@Query("accountId") String accountID);

    /**
     * 获取用户下的医生列表
     *
     * @param accountId
     * @return
     */
    @GET("api/v1/user/doctors")
    Call<List<DoctorInfo>> getDoctorList(@Query("accountId") int accountId);

    /**
     * 获取医生信息
     */
    @GET("api/v1/doctorinfo")
    Call<DoctorInfo> getDoctorInfo(@Query("accountId") int accountId);

    @GET("api/v1/healthdevice/type/bind")
    Call<List<Devices>> getDeviceList(@Query("mac") String mac);

    @GET("api/v1/wechat/subscribe")
    Call<HasWechatId> checkUserHasWechatId(@Query("appId") String appId, @Query("accountId") String accountId);


    /**
     * 绑定手机号获取验证码 4位
     * @return
     */
    @GET("api/v1/bind/sendcode")
    Call<String> getPhoneCode4(@Query("phone") String phone);

    /**
     * 登录获取验证码 6位
     * @return
     */
    @GET("api/v1/phonecode")
    Call<String> getPhoneCode6(@Query("phone") String phone);


    /*------------------------------------医生端------------------------------------*/
    /**
     * 获取民族列表
     * @return
     */
    @GET("api/v1/consts?parentSortId=1")
    Call<List<MinZu>> getMinZuList();

    /**
     * 获取医生所管理的用户列表
     */
    @GET("api/v1/doctor/{accountId}/user")
    Call<List<UserInfo>> getUserList(@Path("accountId") int accountId);

    /**
     * 获取医生所属的机构
     */
    @GET("api/v1/doctor/{accountId}/company")
    Call<List<JiGou>> getJiGou(@Path("accountId") int accountId);

    /**
     * 获取所有省市区
     */
    @GET("api/v1/allarea")
    Call<List<City>> getCitys(@Query("page") int page, @Query("row") int row);

    /**
     * 解除与该用户的关系
     * @return
     */
    @DELETE("api/v1/doctor/{accountId}/user")
    Call<String> deleteUser(@Path("accountId") int accountId, @Query("userAccountId") int userAccountId);

    /**
     *
     * @param appId
     * @param openId
     * @return
     */
    @GET("api/v1/accountthird/verify/login")
    Call<BeanOfWecaht> verifyThirdAccount(@Query("appId") String appId, @Query("openId") String openId,
                                          @Query("groupId") long groupId, @Query("doctorAccountId") long doctorAccountId);
    @POST("api/v1/accountthird/phone/login")
    Call<BeanOfWecaht> bindPhoneAndLogin(@Body ParamThirdUserAccount param);

    @GET("api/v1/accountthird/type")
    Call<List<String>> getUserBindThidAccount(@Query("accountId ") long appId);

    // 根据accountId查询用户是否关注公众号
    @GET("api/v1/wechat/subscribe")
    Call<String> getUserBindWxPublic(@Query("appId") String appId, @Query("accountId") long accountId);

    @GET("api/v1/wechat/qr")
    Call<String> createWXQr(@Query("appId") String appId, @Query("expireSeconds") long expireSeconds, @Query("sceneStr") String sceneStr);
    //修改手环登录密码
    @PUT("api/v1/bandpwd")
    Call<String> backupOldPwd(@Query("accountId") int accountId , @Query("bandPwd") String bandPwd);
}

File: apilib/src/main/java/com/guider/health/apilib/IGuiderApi.kt
package com.guider.health.apilib

import com.guider.health.apilib.bean.*
import com.guider.health.apilib.model.WechatUserInfo
import okhttp3.MultipartBody
import retrofit2.Call
import retrofit2.http.*

interface IGuiderApi {
    @Multipart
    @POST("upload/file")
    fun uploadFile(@Part requestBody: MultipartBody.Part?): Call<String?>?

    /**
     * 用户名密码登录
     *
     * @param telAreaCode 国际区号
     * @param phone       手机号
     * @param passwd      密码
     * @return
     */
    @POST("api/v1/login/phonewithpasswd")
    fun loginWithPassword(@Query("telAreaCode") telAreaCode: String?,
                          @Query("phone") phone: String?,
                          @Query("passwd") passwd: String?): Call<TokenInfo?>?

    /**
     * 注册
     *
     * @param data 请求体数据
     * @return
     */
    @POST("api/v2/register/phonewithpasswd")
    fun register(@Body data: Any?): Call<TokenInfo?>?

    /**
     * 第三方登录前检测是否绑定手机号, 返回数据带UserInfo
     * @param appId 调取三方登录的appId
     * @param openId sdk返回的openId或者userId
     * @param groupId 先固定传-1
     * @param doctorAccountId 先固定传-1
     * @return
     */
    @GET("api/v1/accountthird/verify/login")
    fun lineVerifyLogin(@Query("appId") appId: String?,
                        @Query("openId") openId: String?,
                        @Query("groupId") groupId: Int,
                        @Query("doctorAccountId") doctorAccountId: Int): Call<ThirdLoginVerifyBean?>?

    /**
     * 第三方账号登录并绑定手机号
     * @param data
     * @return
     */
    @POST("api/v1/accountthird/phone/login")
    fun lineBindLogin(@Body data: Any?): Call<ThirdBindPhoneBean?>?

    /**
     * 微信登录
     * @param paramWeChatUserInfo 请求体body
     * @return
     */
    @POST("api/v2/third/login/wachat/tokeninfo")
    fun weChatLoginToken(@Body paramWeChatUserInfo: WeChatInfo?): Call<BeanOfWeChat?>?

    /**
     * 微信绑定手机号
     * @param phone 手机号
     * @param code 验证码
     * @param wechatAppId 微信的APPID
     * @param sence 实体类
     * @return
     */
    @POST("api/v2/wechat/bind/phone/token")
    fun weChatBindLogin(@Query("phone") phone: String?,
                        @Query("code") code: String?,
                        @Query("wechatAppId") wechatAppId: String?,
                        @Body sence: WechatUserInfo?): Call<TokenInfo?>?

    /**
     * 获取用户信息
     *
     * @return
     */
    @GET("api/v1/userinfo")
    fun getUserInfo(@Query("accountId") accountID: Int): Call<UserInfo?>?

    /**
     * 完善用户信息
     * @param userInfo 用户信息实体类
     * @return
     */
    @PUT("api/v1/usersimpleinfo")
    fun simpUserInfo(@Body userInfo: UserInfo?): Call<String?>?

    /**
     * 编辑用户完整信息（包括地址）
     * @param userInfo 用户信息
     * @return
     */
    @PUT("api/v1/userinfo")
    fun editUserInfo(@Body userInfo: UserInfo?): Call<UserInfo?>

    /**
     * 验证是否绑定设备
     * @param accountId
     * @return
     */
    @GET("api/v1/opdevice/account/check")
    fun checkIsBindDevice(@Query("accountId") accountId: Int): Call<Any>

    /**
     * 绑定登录账号的设备
     * @param accountId 用户ID
     * @param deviceCode 设备号
     * @return
     */
    @GET("api/v1/opdevice/account/bind")
    fun bindDeviceWithAccount(@Query("accountId") accountId: Int,
                              @Query("deviceCode") deviceCode: String):
            Call<Any?>

    /**
     * 解绑登录账号绑定的所属设备
     * @param accountId 用户ID
     * @param deviceCode 设备号
     */
    @DELETE("api/v1/user/{accountId}/device/unbind")
    fun unBindDeviceWithAccount(@Query("accountId") accountId: Int,
                                @Query("deviceCode") deviceCode: String):Call<Any>

    /**
     * 验证设备是否有绑定
     * @param deviceCode 设备的code
     */
    @GET("api/v1/opdevice/device/check")
    fun verifyDeviceBind(@Query("deviceCode") deviceCode: String): Call<String>

    /**
     * 验证手机号是否存在并绑定设备、添加家人
     * @param userGroupId 群组id
     * @param deviceCode 设备code码
     * @param phone 手机号
     * @param relationShip 设备昵称
     */
    @GET("api/v1/opdevice/phone/check")
    fun devicePhoneVerify(@Query("userGroupId") userGroupId: Int,
                          @Query("deviceCode") deviceCode: String,
                          @Query("phone") phone: String,
                          @Query("relationShip") relationShip: String,
                          @Query("telAreaCode") telAreaCode: String?
    ): Call<Any?>

    /**
     * 为指定群组添加用户
     * @param userGroupId 群组ID
     * @param accountId 家人帐号id
     * @param relationShip 家人关系
     */
    @POST("api/v3/group/bind")
    fun groupAddMemberDevice(
            @Query("userGroupId") userGroupId: Int,
            @Query("accountId") accountId: Int,
            @Query("relationShip") relationShip: String,
    ): Call<List<UserInfo>?>

    /**
     * 获取登录用户的群下所有群员
     * @param accountId 登录ID
     */
    @GET("api/v2/group/userinfos")
    fun getGroupBindMember(@Query("accountId") accountId: Int): Call<CheckBindDeviceBean?>

    /**
     * 解除群与用户绑定关系
     * @param userGroupId 群组id
     * @param accountId 用户id
     */
    @DELETE("api/v1/group/remove")
    fun unBindGroupMember(@Query("userGroupId") userGroupId: Int,
                          @Query("accountId") accountId: Int): Call<List<UserInfo>?>

    /**
     * 获取特定类型区域
     * @param areaType 地区的枚举类型
     * @param parentId 父ID
     */
    @GET("api/v1/area")
    fun getAddressCode(@Query("areaType") areaType: AddressType,
                       @Query("parentId") parentId: Int
    ): Call<List<AddressWithCodeBean>>

    /**
     * 手机号注册验证
     * @param telAreaCode 国际区号
     * @param phone 手机号
     */
    @GET("api/v1/register/phone/check")
    fun checkPhoneIsRegister(@Query("telAreaCode") telAreaCode: String?,
                             @Query("phone") phone: String?): Call<String>

    /**
     * 查询设备的电子围栏信息
     * @param deviceCode 设备码
     */
    @GET("api/v1/opdevice/fence")
    fun getElectronicFence(@Query("deviceCode") deviceCode: String?):
            Call<Any>

    /**
     * 设置设备的电子围栏信息
     * @param data 设备码+地址围栏信息的latlng列表
     */
    @POST("api/v1/opdevice/fence")
    fun setElectronicFence(@Body data: Any): Call<String>

    /**
     * 运动轨迹
     * @param accountId 查询用户的id
     * @param page 页数，查全部的时候为-1
     * @param row 当查全部的时候row不生效
     * @param startTime 开始时间
     * @param endTime 结束时间
     */
    @GET("api/v1/userpostion")
    fun userPosition( @Query("accountId") accountId: Int,
                      @Query("page") page: Int,
                      @Query("row") row: Int,
                      @Query("startTime") startTime: String,
                      @Query("endTime") endTime: String):Call<List<UserPositionListBean>>
}